CI: Build on Mac

ResearchDaniel opened the pull request:
Changes proposed in this PR:
 * Add mac to CI

ResearchDaniel committed:
CI: Build on Mac

ResearchDaniel committed:
inviwo.yml indent fix

ResearchDaniel committed:
inviwo.yml indent fix

ResearchDaniel committed:
CI: update brew command

ResearchDaniel committed:
CI: Only do doxygen on windows

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -23,14 +23,14 @@ jobs:
-         #   mono: 'mono'
-         #   cmake: '-DIVW_MODULE_HDF5=OFF'
-         #   generator: 'Ninja'
-         #   targets: 'ALL_BUILD package'
```
Why did you change the target name. It is called ALL_BUILD

ResearchDaniel replied:
No, it is not on Ninja

ResearchDaniel replied:
https://cmake.org/cmake/help/git-stage/guide/user-interaction/index.html#id18

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -23,14 +23,14 @@ jobs:
-         #   triplet: 'x64-osx'
-         #   mono: 'mono'
-         #   cmake: '-DIVW_MODULE_HDF5=OFF'
-         #   generator: 'Ninja'
```
generator is not used anymore. I would remove generator and specify a preset instead.

petersteneteg commented on CMakePresets.json:
```diff
@@ -67,18 +67,25 @@
     },
 
     {
-      "name": "msvc-dev",
```
I don't understand why you have renamed this?
We this preset adds is just the generator, then it should be name accordingly
adding x64 is pointless to use in the name since that is the only value we support. 
We can also remove this preset on other platforms with conditionals (https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#condition) then windows is also not needed. Hence MSVC should be in the name.

ResearchDaniel replied:
Ok, this was to avoid duplicating the cmake setup in the CI script.

petersteneteg replied:
I use them extensively in the CMake GUI, so they should make clean sense there...

ResearchDaniel replied:
Changed it back and moved the cmake preset setup to the matrix cmake config instead.

petersteneteg commented on CMakePresets.json:
```diff
@@ -67,18 +67,25 @@
+      "inherits" : ["vcpkg", "x64-windows-dev"]
+    },
+    {
+      "name": "x64-osx-dev-vcpkg",
```
no point in putting OSX in the name of this preset.

ResearchDaniel replied:
fixed

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -141,12 +149,14 @@ jobs:
         path: ${{ matrix.artifact }}
 
     - name: Upload doxygen
+      if: runner.os != 'Windows'
```
this should be runner.os == Windows. 
but I think it would be better to make this a setting in the build matrix...

ResearchDaniel replied:
I think that it would be better if we moved these steps to separate job(s)

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -116,16 +113,27 @@ jobs:
+        -DVCPKG_TARGET_TRIPLET='${{ matrix.triplet }}'
+        -DCMAKE_BUILD_TYPE=${{ env.BUILD_TYPE }}
+        -DIVW_DOXYGEN_PROJECT=OFF
+        ${{ matrix.cmake }}
```
the Doxygen stuff can be moved into matrix.cmake then you don't have to repeat the step.

ResearchDaniel replied:
Good point, fixed

petersteneteg reviewed (changes requested):
might enable ubuntu to...

ResearchDaniel committed:
CI: Simplify workflow. Enable Linux

ResearchDaniel committed:
CI: Refactor

ResearchDaniel committed:
Assimp: Fix for assimp cmake config not found.

https://github.com/microsoft/vcpkg/pull/14554
Packaging: Remove compressed files when packaging installer to save space for CI

ResearchDaniel committed:
CI: Build Qt on mac to enable support for packaging (testing)

ResearchDaniel committed:
Qt: clang fix

ResearchDaniel committed:
CPack: Use macdeployqt to deal with RPATH and Qt dependencies

CRoaring: Fix RPATH bug
Assimp: Fix IrrXML dependency

ResearchDaniel committed:
CI: Mac, try to use standard Qt

ResearchDaniel committed:
CI: use gcc 11

ResearchDaniel committed:
CI: Test mac installer

ResearchDaniel committed:
CI: workflow renaming

ResearchDaniel committed:
CI: indent fix

ResearchDaniel committed:
CI: Upload MacOS dmg installer on release

ResearchDaniel commented:
Kudos to https://github.com/keepassxreboot/keepassxc/commit/ab153a24ec2d5963192db63076bc0be1afa3ad95# which got me on the right track of running macdeployqt during packaging instead of after building and  https://github.com/openstreetmap/merkaartor/commit/a0404fded32e4e794d91b5c9e198ab8ac08b8c2b for how to setup the CPACK_PRE_BUILD_SCRIPTS do so.

ResearchDaniel committed:
CI: Remove unused code and macdeployqt logging

ResearchDaniel committed:
Inviwo: Remove unused code

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -13,39 +13,38 @@ jobs:
-            generator: 'Visual Studio 17 2022'
+            cmake: '--preset msvc-dev-vcpkg -DIVW_MODULE_HDF5=ON -DIVW_DOXYGEN_PROJECT=ON 
+                    -DIVW_DOXYGEN_LATEX_PATH="C:/tools/TinyTeX/bin/win32/pdflatex.exe"'
+            cmake_preset: 'msvc-dev-vcpkg'
```
this was never used, since you also put in in the cmake variable...

ResearchDaniel replied:
Good catch

petersteneteg commented on apps/inviwo/packaging/packaging.cmake:
```diff
@@ -87,16 +87,29 @@ if(WIN32)
     endif()
-
+    
+    get_target_property(_qmake_executable Qt${QT_VERSION_MAJOR}::qmake IMPORTED_LOCATION)
```
why the odd _ prefix for the names?

ResearchDaniel replied:
Fixed

petersteneteg commented on cmake/deploy-osx.cmake.in:
```diff
@@ -0,0 +1,15 @@
+set(MACDEPLOYQT "@MACDEPLOYQT@")
```
Is this only for adding the qt libs/plugins or does this handle rpaths to? 
What about the rpath for other libs? does cmake fix that with the new settings? 
Did we not have some other code to install the qt libs, is that still active?

ResearchDaniel replied:
> Is this only for adding the qt libs/plugins or does this handle rpaths to?

It copies qt libs/plugins and adjusts the RPATHs to those for all the libraries
> What about the rpath for other libs? does cmake fix that with the new settings?

Yes, with the new settings the RPATH of all binaries are set to @loader_path and @loader_path/../Frameworks when packaged

> Did we not have some other code to install the qt libs, is that still active?

Yes, ivw_deploy_qt and ivw_qt_add_to_install 
**ivw_deploy_qt** is only active on Windows to be able to run inviwo outside of the dev environment. We might want to refactor this.

**ivw_qt_add_to_install** is probably not needed on mac/windows anymore. I have not tried to remove it. On linux we would need to use an open source tool (I think that it si called linuxdeployqt or something).

petersteneteg replied:
https://github.com/inviwo/inviwo/blob/f37dcbdfda16646529283ae007339c3b95a447f8/cmake/installutils.cmake#L279

seems like that might do duplicated work then? should we use the same system on windows?

petersteneteg replied:
might as well use windeployqt on windows, and get rid of that crazy code to find the plugins in ivw_qt_add_to_install?

ResearchDaniel replied:
I am testing it on Mac now but don’t have a windows machine right now. Will have to hope that the CI build captures any errors :)

since we already run windeployqt after each build I think that it should be possible to remove all of the stuff, but otherwise it will probably work if we just run it again during packaging

ResearchDaniel replied:
Works on Windows now as well. You should checkout and test it :)
We might want to filter out which dlls to run windeployqt on, e.g., only inviwo-module dlls... However, it seems to work now so perhaps no need to mess with it more.

petersteneteg commented on cmake/globalconfig.cmake:
```diff
@@ -230,6 +230,28 @@ endif()
+    # can be used on other computers.
+    # See:
+    # https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#always-full-rpath
+    set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
```
does this replace the work that fixup_bundle did before? That did the rpath rewriting before..

ResearchDaniel replied:
I think that fixup_bundle was only done on qt plugins before. But yes, it ensures that the RPATH for installed binaries are adjusted.

ResearchDaniel committed:
Inviwo/CI: Remove unused code and rename CMake variables

ResearchDaniel committed:
CMake: Deprecate ivw_qt_add_to_install and ivw_default_install_comp_targets

xxxdeployqt will take care of copying plugins so ivw_qt_add_to_install is not needed anymore.

ResearchDaniel committed:
Packaging: Run windeployqt during packaging.

ResearchDaniel committed:
Packaging: windeployqt test fix

ResearchDaniel committed:
Packaging: Logging for debugging

ResearchDaniel committed:
Packaging: Run windeployqt on both executables and dll libraries

ResearchDaniel committed:
Core: GCC compile error fix

class template placeholder ‘inviwo::util::span’ not permitted in this context
til::Combinations comb(util::span{bases}, searchGroupSize);

Turned on linux CI

ResearchDaniel committed:
Packaging: disregard failed windeployqt errors

ResearchDaniel committed:
Core: Permutation unit test gcc fix

ResearchDaniel committed:
CI: Disable benchmark on linux until compile error has been fixed.

ResearchDaniel committed:
CImg: GCC compile fix

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89062

ResearchDaniel committed:
CImg: GCC compile fix

ResearchDaniel committed:
Packaging: Added required CPack field for unix debian packaging

ResearchDaniel committed:
Core: Format fix

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -13,39 +13,39 @@ jobs:
     strategy:
       matrix:
-        os: [windows-latest] #, macos-latest, ubuntu-20.04]
+        os: [windows-latest, macos-latest] #, ubuntu-20.04]
```
is linux still failing? seemed like it was successful in the test?

ResearchDaniel replied:
It runs and works due to "include" below. Added this back

petersteneteg commented on .github/workflows/inviwo.yml:
```diff
@@ -89,20 +90,20 @@ jobs:
       shell: bash
       if: runner.os == 'Linux'
       run: |
         sudo apt install libxinerama-dev libxcursor-dev xorg-dev libglu1-mesa-dev libglew-dev ninja-build doxygen-latex
```
could probably skip Doxygen here too

ResearchDaniel replied:
Fixed

petersteneteg reviewed (dismissed):
The windows installer seems to work for me

Did you disable linux again? seems from the CI run that is was active, but it is commented now?

ResearchDaniel committed:
CI: Remove doxygen dependency from linux build

ResearchDaniel commented:
> Did you disable linux again? seems from the CI run that is was active, but it is commented now?

Active due to the include. Added the comment thingy now

ResearchDaniel committed:
CI: .deb Linux package file ending
